src/lib/ams.c

```diff
@@ -0,0 +1,606 @@
+						shared_c = ams_acquire(ams->ams_context->shared);
+						shared_c->slot_uses[slot]--;
+						shared_c->slot_done[slot] |= BIT(ams_target.consumer_core_id);
+						ams_release(shared_c);
+					}
+				}
+			} else {
+				/* message already forwarded, nothing to do here */
+				ams_release(shared_c);
+			}
+		}
+
+		/* acquire shared context lock again */
+		shared_c = ams_acquire(ams->ams_context->shared);
+	}
+
+	if (incoming)
+		shared_c->slot_uses[incoming_slot]--;
+
+	ams_release(shared_c);
+
+	if (!found_any)
+		tr_err(&ams_tr, "No entries found!");
+
+	return err;
+}
+
+int ams_send(const struct ams_message_payload *const ams_message_payload)
+{
+	struct async_message_service *ams = *arch_ams_get();
+
+	return ams_message_send_internal(ams, ams_message_payload, AMS_ANY_ID, AMS_ANY_ID,
+					 AMS_INVALID_SLOT);
+}
+
+int ams_message_send_mi(struct async_message_service *ams,
+			const struct ams_message_payload *const ams_message_payload,
+			uint16_t target_module, uint16_t target_instance)
+{
+	return ams_message_send_internal(ams, ams_message_payload, target_module,
+					 target_instance, AMS_INVALID_SLOT);
+}
+
+int ams_send_mi(const struct ams_message_payload *const ams_message_payload,
+		uint16_t module_id, uint16_t instance_id)
+{
+	struct async_message_service *ams = *arch_ams_get();
+
+	return ams_message_send_mi(ams, ams_message_payload, module_id, instance_id);
+}
+
+static int ams_process_slot(struct async_message_service *ams, uint32_t slot)
+{
+	struct ams_shared_context __sparse_cache *shared_c;
+	struct ams_message_payload msg;
+	uint16_t module_id;
+	uint16_t instance_id;
+
+	shared_c = ams_acquire(ams->ams_context->shared);
+
+	msg = shared_c->slots[slot].u.msg;
+	module_id = shared_c->slots[slot].module_id;
+	instance_id = shared_c->slots[slot].instance_id;
+
+	ams_release(shared_c);
+	tr_info(&ams_tr, "ams_process_slot slot %d msg %d from 0x%08x",
+		slot, msg.message_type_id,
+		msg.producer_module_id << 16 | msg.producer_instance_id);
+
+	return ams_message_send_internal(ams, &msg, module_id, instance_id, slot);
+}
+
+#if CONFIG_SMP
+
+static void ams_task_add_slot_to_process(struct ams_task *ams_task, uint32_t slot)
+{
+	int flags;
+
+	irq_local_disable(flags);
+	ams_task->pending_slots |= BIT(slot);
+	irq_local_enable(flags);
+}
+
+int process_incoming_message(uint32_t slot)
+{
+	struct async_message_service *ams = *arch_ams_get();
+	struct ams_task *task = &ams->ams_task;
+
+	ams_task_add_slot_to_process(task, slot);
+
+	return schedule_task(&task->ams_task, 0, 10000);
+}
+
+#endif /* CONFIG_SMP */
+
+/* ams task */
+
+static enum task_state process_message(void *arg)
+{
+	struct ams_task *ams_task = arg;
+	uint32_t slot;
+	int flags;
+
+	if (ams_task->pending_slots == 0) {
+		tr_err(&ams_tr, "Could not process message! Skipping.");
+		return SOF_TASK_STATE_COMPLETED;
+	}
+
+	#ifdef __XCC__
+	slot = 31 - XT_NSAU(ams_task->pending_slots);
+	#else
+	slot = 31 - clz(ams_task->pending_slots);
+	#endif
```

Comment:
there should be a static inline function for this, having compiler-dependent stuff in the middle of a long file is not a good practice